### PR TITLE
Added Tensorflow support to MLDB (MLDB-1203)

### PR DESCRIPTION
**Pull request by jeremybarnes**
Allows Tensorflow graphs to be loaded and executed by MLDB.  There are a couple of small cleanups before the code is merged, but this is the vast majority of the functionality.

### CODE REVIEW

File: sql/builtin_functions.cc [hunk 1]
```diff
@@ -1905,6 +1905,39 @@ BoundFunction horizontal_max(const std::vector<BoundSqlExpression> & args)
...
+                                   const Coord & prefix,
+                                   const CellValue & val,
+                                   Date atomTs)
+                    {
+                        if (!val.empty()) {
+                            double curr = val.toDouble();
+                            if(std::isnan(max_val) || curr > max_val) {
+                                ts = atomTs;
+                                max_val = curr;
+                                max_key = prefix.toString() + "." + columnName.toString();
```
Comment (guyd) [inline sql/builtin_functions.cc]: May be we can capture the prefix and column name separately and perform the string operation outside the loop.

File: sql/builtin_functions.cc [hunk 2]
```diff
@@ -2169,6 +2202,41 @@ BoundFunction upper(const std::vector<BoundSqlExpression> & args)
 
 static RegisterBuiltin registerUpper(upper, "upper");
 
+BoundFunction flatten(const std::vector<BoundSqlExpression> & args)
+{
+    // Return an expression but with the timestamp modified to something else
+
+    if (args.size() != 1)
```
Comment (guyd) [inline sql/builtin_functions.cc]: We have checkArgsSize for that.

File: sql/expression_value.cc [hunk 3]
```diff
@@ -1802,6 +1802,33 @@ getEmbeddingShape() const
...
+        if (rowLength() > totalLength) {
+            throw HttpReturnException(400, "Attempt to enlarge embedding by resizing");
+        }
+        return ExpressionValue::embedding(ts_, embedding_->data_,
+                                          embedding_->storageType_,
+                                          std::move(newShape));
+    }
+    }
+
+    throw HttpReturnException(500, "Unknown storage type for getEmbeddingShape()");
```
Comment (guyd) [inline sql/expression_value.cc]: getEmbeddingShape -> reshape

File: tensorflow/doc/TensorflowGraph.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+![](%%config function tensorflow.graph)
+
+## Example
+
+The following function configuration will load the Tensorflow "inception"
+model (which is made publicly available by Google) to predict the top 5
+classes of an arbitrary image passed in as a URL.
+
+```
```
Comment (guyd) [inline tensorflow/doc/TensorflowGraph.md]: I guess you will complete that in a subsequent commit.

File: tensorflow/tensorflow_plugin.cc [hunk 5]
```diff
@@ -0,0 +1,1182 @@
...
+
+DECLARE_STRUCTURE_DESCRIPTION(TensorflowGraphConfig);
+
+DEFINE_STRUCTURE_DESCRIPTION(TensorflowGraphConfig);
+
+TensorflowGraphConfigDescription::
+TensorflowGraphConfigDescription()
+{
+    addField("modelFileUrl", &TensorflowGraphConfig::modelFileUrl,
+             "Model file to load graph from.  This is probable a .pb "
```
Comment (guyd) [inline tensorflow/tensorflow_plugin.cc]: probable -> probably

File: sql/builtin_functions.cc (hunk 2 above)
Comment (steadtler) [inline sql/builtin_functions.cc]: The catch lower in the call stack will print the function name, by the way, so we dont need to repeat it.

**Comment by guyd** [comment]
Reviewed briefly the tensor flow function.  Seems ok beside a few typos.  +1

File: tensorflow/doc/TensorflowGraph.md (hunk 4 above)
Comment (jeremybarnes) [inline tensorflow/doc/TensorflowGraph.md]: I completed it :)

**Comment by jeremybarnes** [comment]
Based on conversation with @nicolaskruchten, this will be merged by him at a time that makes sense with respect to the release schedule.

**Pull request merged**